### PR TITLE
Remodel component interfaces

### PR DESCRIPTION
**Pull request by Karsten1987**
Second part of #164 

Introduces `StateHandle` and `CommandHandle`. 
Adapts the hardware components interfaces to instantiate these handles given the configuration info passed to the components on a prior call to `configure`.

Sits on top of #201

### CODE REVIEW

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (bmagyar) [inline hardware_interface/include/hardware_interface/handle.hpp:35]: This feels highly redundant. You changed a pass-by-const-ref to pass-by-value + std::move. I'd really prefer if we didn't abuse `std::move` for petty stuff like this.

Comment (Karsten1987) [reply hardware_interface/include/hardware_interface/handle.hpp:35]: I think that's a really valid case especially given that these names are mainly short strings, so SSO kicks in perfectly.

File: hardware_interface/test/test_component_interfaces.cpp
Comment (bmagyar) [inline hardware_interface/test/test_component_interfaces.cpp:154]: not really "actuator info", I suggest updating these comments to just use "info"

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/handle.hpp:35]: This is superstitious programming and premature optimization. I trust to compiler to know where to optimize and I am still strongly for keeping the code simple. 

To back up my claim, I've ran google benchmark on this. The code can be found [here](https://gist.github.com/bmagyar/d3771595e7567cb00479a76b3a711cdb). Took the long names from our urdf examples page.
These benchmarks show that the `std::move` alternative (marked with `/0`) has a few nanoseconds of overhead compared to the other one both in debug and release mode. Even if it was the opposite way, I'd argue to keep the code simpler and swallow that 3ns overhead per handle and *avoid std::move*.

```
***WARNING*** Library was built as DEBUG. Timings may be affected.
-----------------------------------------------------------------------------------------
Benchmark                                              Time            CPU       Iterations
------------------------------------------------------------------------------------------
MyFixture<StateHandle>/ShortNames_StateHandle/0       69.0 ns         69.0 ns     10294137
MyFixture<StateHandle>/ShortNames_StateHandle/1       53.5 ns         53.5 ns     12842977
MyFixture<StateHandle>/LongNames_StateHandle/0        87.1 ns         87.1 ns      8088183
MyFixture<StateHandle>/LongNames_StateHandle/1        76.3 ns         76.3 ns      9228388
```
The same in a release build:
```
------------------------------------------------------------------------------------------
Benchmark                                              Time           CPU      Iterations
------------------------------------------------------------------------------------------
MyFixture<StateHandle>/ShortNames_StateHandle/0       17.3 ns         17.3 ns     40833624
MyFixture<StateHandle>/ShortNames_StateHandle/1       13.7 ns         13.7 ns     51205075
MyFixture<StateHandle>/LongNames_StateHandle/0        30.9 ns         30.9 ns     21597366
MyFixture<StateHandle>/LongNames_StateHandle/1        27.5 ns         27.5 ns     25822783
```

((( I made the fixture templated on the handle type but realized that there was no point in testing multiple, sorry for the added complexity)))

**Review by bmagyar — CHANGES_REQUESTED** [review]
Looks good in general. Happy to merge once we settled the std::move point.

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (Karsten1987) [reply hardware_interface/include/hardware_interface/handle.hpp:35]: I think that's a real common use-case. https://www.bfilipek.com/2018/08/init-string-member.html
It doesn't make a big difference when dealing with SSO, you're right.

File: hardware_interface/test/test_component_interfaces.cpp
Comment (Karsten1987) [reply hardware_interface/test/test_component_interfaces.cpp:154]: it's been like that in the original PR already.

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/handle.hpp:35]: Please get rid of `std::move`. This constructor is now also inconsistent with `ReadWriteHandle`.

**Review by bmagyar — CHANGES_REQUESTED** [review]

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (Karsten1987) [reply hardware_interface/include/hardware_interface/handle.hpp]: done in a36352e

File: hardware_interface/test/test_component_interfaces.cpp
Comment (Karsten1987) [reply hardware_interface/test/test_component_interfaces.cpp]: done in 5a27bf8

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (destogl) [reply hardware_interface/include/hardware_interface/handle.hpp:89]: Since we are using inheritance, this is kind of overloaded with the templates. Am I missing something?

File: hardware_interface/include/hardware_interface/hardware_info.hpp
Comment (destogl) [inline hardware_interface/include/hardware_interface/hardware_info.hpp:55]: For the sensors one wants to use links sometimes, e.g., `robot_tcp` for an FTS
```suggestion
   * \brief map of sensors provided by the hardware where the key is the joint or link name.
```

File: hardware_interface/include/hardware_interface/joint_handle.hpp
Comment (destogl) [inline hardware_interface/include/hardware_interface/joint_handle.hpp:23]: Why not use `CommandHandle`

Comment (destogl) [reply hardware_interface/include/hardware_interface/joint_handle.hpp:23]: Or is this also part of the transmissions?

**Review by destogl — COMMENTED** [review]
I added few discussion points.

Will you rename "Handle" --> "Interface", as discussed in an issue (i can not find it right now...)

File: hardware_interface/include/hardware_interface/actuator_handle.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/actuator_handle.hpp:25]: Yes this is for handling transmissions. We don't want `CommandHandle` to have any relation to `ActuatorHandle` but in the implementation.

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/handle.hpp:89]: This is a mixin implementation. It's an elegant way of injecting code without creating a subtype and having to worry about visibility exposure (private or protected inheritance would cover the no-subtype requirement). It also leaves the option open for us to reimplement these classes internally. The mixin usage here basically creates a "copy paste" effect.

File: hardware_interface/include/hardware_interface/hardware_info.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/hardware_info.hpp:55]: Good catch

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/handle.hpp:124]: Not for the majority of use cases. This was added because the joint limits implementation required it

File: hardware_interface/include/hardware_interface/components/system_interface.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/components/system_interface.hpp:71]: This is written approximately 129387 times in this PR. I suggest we move the discussion on this to a new issue and let the PR proceed.

File: hardware_interface/include/hardware_interface/joint_handle.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/joint_handle.hpp:23]: This PR is a wee bit misleading in the title and description. Not only component interfaces were remodeled but the existing ones were also touched up to avoid code duplication.

**Comment by bmagyar** [comment]
I've rebased this PR after merging #209

File: hardware_interface/include/hardware_interface/actuator_handle.hpp
Comment (destogl) [reply hardware_interface/include/hardware_interface/actuator_handle.hpp:25]: Maybe we should rename this so that the nomenclature do not collide with components.

"Joint" is OK because is not used anymore, we should check for the "Actuator"

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (destogl) [reply hardware_interface/include/hardware_interface/handle.hpp]: Should we remove this in the future? I don't think this is sensible.

File: hardware_interface/include/hardware_interface/actuator_handle.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/actuator_handle.hpp:25]: In a few PR-s time, they should be removed. Let's save time on that renaming ;)

File: hardware_interface/include/hardware_interface/handle.hpp
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/handle.hpp]: Let's open a new thread for this. I agree that some sort of clarity is necessary on this as why we'd have it for name and not for the interface, how'd you have the same thing where you set value *and* interface, it'd clash with this one at the name level, etc.

**Review by bmagyar — APPROVED** [review]

**Pull request merged**
